src/utils/thread.zig

```diff
@@ -7,41 +7,64 @@ const Mutex = std.Thread.Mutex;
 const ThreadPool = @import("../sync/thread_pool.zig").ThreadPool;
 const Batch = ThreadPool.Batch;
 
-/// Spawns tasks and outputs the list of handles for the spawned threads.
-/// Task function should accept `params ++ .{ start_index, end_index, thread_id }` as its parameter tuple.
+pub const TaskParams = struct {
+    start_index: usize,
+    end_index: usize,
+    thread_id: usize,
+};
+
+fn chunkSizeAndThreadCount(data_len: usize, max_n_threads: usize) struct { usize, usize } {
+    var chunk_size = data_len / max_n_threads;
+    var n_threads = max_n_threads;
+    if (chunk_size == 0) {
+        n_threads = 1;
+        chunk_size = data_len;
+    }
+    return .{ chunk_size, n_threads };
+}
+
 pub fn spawnThreadTasks(
-    /// This list is cleared, and then filled with the handles for the spawned task threads.
-    /// On successful call, all threads were appropriately spawned.
-    handles: *std.ArrayList(std.Thread),
-    comptime taskFn: anytype,
-    params: anytype,
+    wait_group: *std.Thread.WaitGroup,
     data_len: usize,
     max_n_threads: usize,
-) (std.mem.Allocator.Error || std.Thread.SpawnError)!void {
-    const chunk_size, const n_threads = blk: {
-        var chunk_size = data_len / max_n_threads;
-        var n_threads = max_n_threads;
-        if (chunk_size == 0) {
-            n_threads = 1;
-            chunk_size = data_len;
+    /// If non-null, set to the coverage over the data which was achieved.
+    /// On a successful call, this will be equal to `data_len`.
+    /// On a failed call, this will be less than `data_len`,
+    /// representing the length of the data which was successfully
+    maybe_coverage: ?*usize,
+    comptime taskFn: anytype,
+    params: anytype,
+) std.Thread.SpawnError!void {
+    const chunk_size, const n_threads = chunkSizeAndThreadCount(data_len, max_n_threads);
+
+    if (maybe_coverage) |coverage| coverage.* = 0;
+
+    const S = struct {
+        fn taskFnWg(wg: *std.Thread.WaitGroup, fn_params: @TypeOf(params), task_params: TaskParams) @typeInfo(@TypeOf(taskFn)).Fn.return_type.? {
```

Comment:
Already exists in the stdlib in the form of `wg.spawnManager`. I hate that it exists but I mean, might as well use it.

Reply:
To add, I don't see a point to catching `spawn` failing. It would only fail under the most extreme conditions, it wouldn't be something that we could recover from.

Reply:
The caller could pretty reasonably recover by doing the remaining work on their own thread, if it were something critical. I get that ideally we move away from spawnThreadTasks to something better, but insolong as it is here, I don't see the harm in handling edge cases.

Reply:
The good thing about `spawnManager` is that it can't fail, and just gives ownership of the task to the caller thread to ensure it finishes. That would be a slowdown of course, but it would ensure that what needs to be done gets done.

Reply:
I don't see the need to pay that cost in cases where the work isn't critical - this implementation lets the caller ignore it if that is the case.

Reply:
`spawn` failing is not something we should really be trying to optimize. The "cost" we pay as you say is that either the caller needs to explicitly handle setting up logic for re-running threads, which will almost certainly just fail again (maybe resulting in a loop, so then we need to add extra retry attempts logic), or the function call "just works" and it may have taken some extra time but the work gets done anyways. 

`spawn` return an error is not an edge case, it's a large problem and recovering from it is highly unlikely.

Reply:
I understand the idea now, sorry it took me a couple times to get it :). This is fine for me and better than the stdlib one.